### PR TITLE
Only probe a small allowlist of loop devices

### PR DESCRIPTION
**Pull request by DemiMarie**
Others might hold attacker-controlled data, such as TemplateVM or
StandaloneVM package caches.

Fixes QubesOS/qubes-issues#6355

### CODE REVIEW

File: system-config/skip-encrypted-swap.rules
Comment (marmarek) [inline system-config/skip-encrypted-swap.rules:13]: This assumes a specific device name, but such device is not created anywhere in the qubes code base, nor this assumed name is documented anywhere. Are you going to submit a separate PR that configures encrypted swap?

File: system-config/00-qubes-ignore-devices.rules
Comment (marmarek) [inline system-config/00-qubes-ignore-devices.rules:5]: Same question about vm-kernels dir.

**Review by marmarek — CHANGES_REQUESTED** [review]
As you noted elsewhere, this likely breaks qvm-block attaching loop devices made by the user. Is that correct? If so, either qvm-block related udev rules needs an adjustment to list such devices (note you don't really need udev device parsing there, as for the loop devices we use just file path as in a description, not any fs or device metadata or such like for physical devices). Alternatively, we need a tool (or an option to qvm-block?) to explicitly include specific file/loop device in the list.

Besides the specific comments, some general remarks (applicable for other, including future, PRs too):
1. package-managed files should be installed into system directories like `/lib`, `/usr` etc whenever possible. `/etc` should be left for the user modifications. It isn't always possible, but in the great majority of the cases, it is. This avoid all kind of issues with package/user conflicts - like reverting explicit user change, or not applying system update (because the new file ends up in an `.rpmnew` file).
2. Avoid implicit cross-package dependencies when reasonably possible. Every such dependency makes updating more complex and fragile (cases like failures depending on package installation order, or updating just some packages). Sooner or later (rather sooner) such implicit dependencies need to be named explicitly (in package metadata), but it still requires mental load to maintain them (especially when given commit is a backport candidate). In context of this PR: a file that needs an inclusion in initramfs, should have matching dracut config/module in the same package as the file itself.

File: system-config/skip-encrypted-swap.rules
Comment (DemiMarie) [reply system-config/skip-encrypted-swap.rules:9]: I believe LVM volumes are covered by the `DM_LV_NAME` case.

These rules all originated from a very specific use case on my personal laptop, so it is not surprising that some cleanup is needed.

File: system-config/skip-encrypted-swap.rules
Comment (DemiMarie) [reply system-config/skip-encrypted-swap.rules:13]: Indeed I will.  I should probably split the parts needed for encrypted swap into separate PRs.

File: system-config/skip-encrypted-swap.rules
Comment (DemiMarie) [reply system-config/skip-encrypted-swap.rules:6]: Not sure; I wrote this code back in early 2020 when I was much less familiar with QubesOS internals.

File: system-config/skip-encrypted-swap.rules
Comment (DemiMarie) [reply system-config/skip-encrypted-swap.rules:1]: These are separate for a reason, but `skip-encrypted-swap.rules` needs to come very late.  I can prefix it with `99z-`.

File: dracut/modules.d/90qubes-udev/module-setup.sh
Comment (DemiMarie) [reply dracut/modules.d/90qubes-udev/module-setup.sh:5]: This is another artifact that is specific to my personal installation :smile:.

File: dracut/modules.d/90qubes-udev/module-setup.sh
Comment (DemiMarie) [reply dracut/modules.d/90qubes-udev/module-setup.sh:4]: Indeed.  The use of `/etc` here is *also* historical: these files are in my *personal* `/etc`, and I forgot to rename them :laughing:.

**Review by DemiMarie — COMMENTED** [review]
Good catch @marmarek!

**Comment by marmarek** [comment]
A minor comment: can you not push a merge master commit into a PR unless there is some specific need (conflict resolution, dependency on some newer stuff in master etc)? All our CI will merge it with master (or whatever target branch is) for testing anyway. And if you need some newer stuff from master branch, rebase is more preferable is not too hard. The goal is to keep git history as clean as practically possible (in fact, the main and perhaps the only reason we merge PRs via merge commits and not rebase is to preserve commit signatures).

File: system-config/skip-encrypted-swap.rules
Comment (marmarek) [reply system-config/skip-encrypted-swap.rules:9]: Still, the `ENV{DM_NAME}=="vm-*"` for LVM will match only volume _groups_ starting with `vm-`. Oh I see github mangled my comment above, fixed.

**Comment by marmarek** [comment]
Build failed:
```
DEBUG: RPM build errors:
DEBUG: error: File must begin with "/": %_udevrulesdir/00-qubes-ignore-devices.rules
DEBUG: error: File must begin with "/": %_udevrulesdir/12-qubes-ignore-lvm-devices.rules
DEBUG: error: File must begin with "/": %_presetdir/75-qubes-dom0.preset
DEBUG: error: File must begin with "/": %_presetdir/99-qubes-default-disable.preset
DEBUG: error: File must begin with "/": %_userpresetdir/75-qubes-dom0-user.preset
DEBUG:     line 69: It's not recommended to have unversioned Obsoletes: Obsoletes: os-prober
DEBUG:     File must begin with "/": %_udevrulesdir/00-qubes-ignore-devices.rules
DEBUG:     File must begin with "/": %_udevrulesdir/12-qubes-ignore-lvm-devices.rules
DEBUG:     File must begin with "/": %_presetdir/75-qubes-dom0.preset
DEBUG:     File must begin with "/": %_presetdir/99-qubes-default-disable.preset
DEBUG:     File must begin with "/": %_userpresetdir/75-qubes-dom0-user.preset
DEBUG: Child return code was: 1
```

**Comment by marmarek** [comment]
> system_tests_basic_vm_qrexec_gui
> 
>     * TC_00_Basic: [test_120_start_standalone_with_cdrom_dom0](https://openqa.qubes-os.org/tests/16387#step/TC_00_Basic/4)

So, this (unsurprisingly) breaks `qvm-start --cdrom dom0:/some/file/path`. Specific error:

> AssertionError: 1 != 0 : b'Timeout waiting for dom0:loop4 device to appear\n'

Udev should not look at the content of such loop devices (which I think you achieved here), but qvm-block still should list them (only standard VM disks should be excluded). I think you need to adjust qvm-block related udev rules (in linux-utils repo).

**Comment by DemiMarie** [comment]
> > system_tests_basic_vm_qrexec_gui
> > ```
> > * TC_00_Basic: [test_120_start_standalone_with_cdrom_dom0](https://openqa.qubes-os.org/tests/16387#step/TC_00_Basic/4)
> > ```
> 
> So, this (unsurprisingly) breaks `qvm-start --cdrom dom0:/some/file/path`. Specific error:
> 
> > AssertionError: 1 != 0 : b'Timeout waiting for dom0:loop4 device to appear\n'
> 
> Udev should not look at the content of such loop devices (which I think you achieved here), but qvm-block still should list them (only standard VM disks should be excluded). I think you need to adjust qvm-block related udev rules (in linux-utils repo).

Would it be better to use a custom script that both created and registered the loop device?

**Comment by marmarek** [comment]
Yes, I think that's a good idea. A qrexec service maybe (more relevant for VM case, but consistent solution for dom0 would be nice too)? That would solve a rather inelegant direct `losetup` call. Such qrexec service could wait until device is created and information set in qubesdb, which should avoid the need for the while-loop waiting for the device.

This will be basically a "give me access to an arbitrary file" service (as long as the caller can also attach a block device), but I don't have much better idea. For some cases (the GUI equivalent - `qubes-vm-boot-from-device`) it could be combined with asking for the file in the VM itself (instead of getting the path from the caller), but I think we need the non-GUI version too.

**Comment by marmarek** [comment]
>     * TC_00_Basic: [test_120_start_standalone_with_cdrom_dom0](https://openqa.qubes-os.org/tests/16928#step/TC_00_Basic/4)

This failure is definitely related to this PR. This is kid of related to https://github.com/QubesOS/qubes-core-admin-linux/pull/71#pullrequestreview-585031330



> * TC_05_StandaloneVM_debian-10: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_debian-10/3)
> 
> * TC_05_StandaloneVM_whonix-gw-15: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_whonix-gw-15/3)
> 
> * TC_05_StandaloneVM_whonix-ws-15: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_whonix-ws-15/3)

Those I'm not sure, but may be related to this change as a whole. The specific error is *within the VM*:
```
resize2fs 1.44.5 (15-Dec-2018)
open: No such file or directory while opening /dev/mapper/dmroot
```
The `/dev/mapper/dmroot` symlink is made by udev normally (it points to `/dev/xvda3` directly in case of writable root volume; in case of read-only root volume, it indeed is a dm device (dm-snapshot built from root+volatile)).

**Comment by DemiMarie** [comment]
> > ```
> > * TC_00_Basic: [test_120_start_standalone_with_cdrom_dom0](https://openqa.qubes-os.org/tests/16928#step/TC_00_Basic/4)
> > ```
> 
> This failure is definitely related to this PR. This is kind of related to [#71 (review)](https://github.com/QubesOS/qubes-core-admin-linux/pull/71#pullrequestreview-585031330)

I am not surprised; I got distracted by the RPM mess and didn’t pay much attention to other stuff.  I have a script that can solve this problem, but it is subject to races against manual `losetup` calls from within the qube; this can be solved by having `xl devd` hold a file descriptor to the device in question.  Nevertheless, this script is no worse than the current situation.

> > * TC_05_StandaloneVM_debian-10: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_debian-10/3)
> > * TC_05_StandaloneVM_whonix-gw-15: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_whonix-gw-15/3)
> > * TC_05_StandaloneVM_whonix-ws-15: [test_101_resize_root_img_online](https://openqa.qubes-os.org/tests/16928#step/TC_05_StandaloneVM_whonix-ws-15/3)
> 
> Those I'm not sure, but may be related to this change as a whole. The specific error is _within the VM_:
> 
> ```
> resize2fs 1.44.5 (15-Dec-2018)
> open: No such file or directory while opening /dev/mapper/dmroot
> ```
> 
> The `/dev/mapper/dmroot` symlink is made by udev normally (it points to `/dev/xvda3` directly in case of writable root volume; in case of read-only root volume, it indeed is a dm device (dm-snapshot built from root+volatile)).

This very much could be the problem.  I will look at it tomorrow.

**Comment by DemiMarie** [comment]
@marmarek will this be ready to merge once QubesOS/qubes-linux-utils#76 lands?

File: system-config/99z-qubes-mark-ready.rules
Comment (marmarek) [inline system-config/99z-qubes-mark-ready.rules:9]: This belongs to your own custom setup, not anything done by standard qubes code, right? This matches raw device mapper name, not LV name (which is covered earlier) and we don't use dm devices named `vm-*`. Am I missing something? Otherwise, plase remove.

File: dracut/modules.d/90qubes-udev/module-setup.sh
Comment (marmarek) [inline dracut/modules.d/90qubes-udev/module-setup.sh:7]: You forgot `rules.d` in the path.

File: dracut/modules.d/90qubes-udev/module-setup.sh
Comment (marmarek) [inline dracut/modules.d/90qubes-udev/module-setup.sh:6]: Regardless of the comment below, this file doesn't exist.

**Review by marmarek — CHANGES_REQUESTED** [review]

**Review by marmarek — APPROVED** [review]

**Pull request merged**
